File: java/lib/src/main/java/com/svix/models/DestinationPatchConfig.java
// This file is @generated
package com.svix.models;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.svix.Utils;

import lombok.*;

import java.util.HashMap;
import java.util.Map;

@ToString
@EqualsAndHashCode
public abstract class DestinationPatchConfig {
    @JsonIgnore
    public String getVariantName() {
        VariantName annotation = this.getClass().getAnnotation(VariantName.class);
        return annotation != null ? annotation.value() : null;
    }

    public abstract JsonNode toJsonNode();

    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("pollingEndpoint")
    public static class PollingEndpoint extends DestinationPatchConfig {
        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().createObjectNode();
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("azureBlobStorage")
    public static class AzureBlobStorage extends DestinationPatchConfig {
        private final AzureBlobStorageConfigPatch azureBlobStorage;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(azureBlobStorage);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("otelTracing")
    public static class OtelTracing extends DestinationPatchConfig {
        private final OtelTracingConfigPatch otelTracing;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(otelTracing);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("fifoEndpoint")
    public static class FifoEndpoint extends DestinationPatchConfig {
        private final SinkHttpConfigPatch fifoEndpoint;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(fifoEndpoint);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("amazonS3")
    public static class AmazonS3 extends DestinationPatchConfig {
        private final S3ConfigPatch amazonS3;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(amazonS3);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("googleCloudStorage")
    public static class GoogleCloudStorage extends DestinationPatchConfig {
        private final GoogleCloudStorageConfigPatch googleCloudStorage;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(googleCloudStorage);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("googleCloudPubSub")
    public static class GoogleCloudPubSub extends DestinationPatchConfig {
        private final GoogleCloudPubSubConfigPatch googleCloudPubSub;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(googleCloudPubSub);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("sqs")
    public static class Sqs extends DestinationPatchConfig {
        private final SqsConfigPatch sqs;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(sqs);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("sns")
    public static class Sns extends DestinationPatchConfig {
        private final SnsConfigPatch sns;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(sns);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("bigQuery")
    public static class BigQuery extends DestinationPatchConfig {
        private final BigQueryConfigPatch bigQuery;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(bigQuery);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("clickhouse")
    public static class Clickhouse extends DestinationPatchConfig {
        private final ClickhouseConfigPatch clickhouse;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(clickhouse);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("eventBridge")
    public static class EventBridge extends DestinationPatchConfig {
        private final EventBridgeConfigPatch eventBridge;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(eventBridge);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("snowflake")
    public static class Snowflake extends DestinationPatchConfig {
        private final SnowflakeConfigPatch snowflake;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(snowflake);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("rabbitMq")
    public static class RabbitMq extends DestinationPatchConfig {
        private final RabbitMqConfigPatch rabbitMq;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(rabbitMq);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("redshift")
    public static class Redshift extends DestinationPatchConfig {
        private final RedshiftConfigPatch redshift;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(redshift);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("postgres")
    public static class Postgres extends DestinationPatchConfig {
        private final PostgresConfigPatch postgres;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(postgres);
        }
    }

    @FunctionalInterface
    private interface TypeFactory {
        DestinationPatchConfig create(JsonNode config);
    }

    private static final Map<String, TypeFactory> TY_M = new HashMap<>();
    private static final ObjectMapper m = Utils.getObjectMapper();

    static {
        TY_M.put("pollingEndpoint", c -> new PollingEndpoint());
        TY_M.put(
                "azureBlobStorage",
                c -> new AzureBlobStorage(m.convertValue(c, AzureBlobStorageConfigPatch.class)));
        TY_M.put(
                "otelTracing",
                c -> new OtelTracing(m.convertValue(c, OtelTracingConfigPatch.class)));
        TY_M.put(
                "fifoEndpoint",
                c -> new FifoEndpoint(m.convertValue(c, SinkHttpConfigPatch.class)));
        TY_M.put("amazonS3", c -> new AmazonS3(m.convertValue(c, S3ConfigPatch.class)));
        TY_M.put(
                "googleCloudStorage",
                c ->
                        new GoogleCloudStorage(
                                m.convertValue(c, GoogleCloudStorageConfigPatch.class)));
        TY_M.put(
                "googleCloudPubSub",
                c -> new GoogleCloudPubSub(m.convertValue(c, GoogleCloudPubSubConfigPatch.class)));
        TY_M.put("sqs", c -> new Sqs(m.convertValue(c, SqsConfigPatch.class)));
        TY_M.put("sns", c -> new Sns(m.convertValue(c, SnsConfigPatch.class)));
        TY_M.put("bigQuery", c -> new BigQuery(m.convertValue(c, BigQueryConfigPatch.class)));
        TY_M.put("clickhouse", c -> new Clickhouse(m.convertValue(c, ClickhouseConfigPatch.class)));
        TY_M.put(
                "eventBridge",
                c -> new EventBridge(m.convertValue(c, EventBridgeConfigPatch.class)));
        TY_M.put("snowflake", c -> new Snowflake(m.convertValue(c, SnowflakeConfigPatch.class)));
        TY_M.put("rabbitMq", c -> new RabbitMq(m.convertValue(c, RabbitMqConfigPatch.class)));
        TY_M.put("redshift", c -> new Redshift(m.convertValue(c, RedshiftConfigPatch.class)));
        TY_M.put("postgres", c -> new Postgres(m.convertValue(c, PostgresConfigPatch.class)));
    }

    public static DestinationPatchConfig fromTypeAndConfig(String type, JsonNode config) {
        TypeFactory factory = TY_M.get(type);
        if (factory == null) {
            throw new IllegalArgumentException("Unknown type: " + type);
        }
        return factory.create(config);
    }
}

File: php/src/Models/DestinationPatchConfig.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Models;

abstract class DestinationPatchConfig implements \JsonSerializable
{
    abstract public function variantName(): string;

    abstract public function configPayload(): mixed;

    public static function pollingEndpoint(): DestinationPatchConfig\PollingEndpoint
    {
        return new DestinationPatchConfig\PollingEndpoint();
    }

    public static function azureBlobStorage(AzureBlobStorageConfigPatch $azureBlobStorage): DestinationPatchConfig\AzureBlobStorage
    {
        return new DestinationPatchConfig\AzureBlobStorage($azureBlobStorage);
    }

    public static function otelTracing(OtelTracingConfigPatch $otelTracing): DestinationPatchConfig\OtelTracing
    {
        return new DestinationPatchConfig\OtelTracing($otelTracing);
    }

    public static function fifoEndpoint(SinkHttpConfigPatch $fifoEndpoint): DestinationPatchConfig\FifoEndpoint
    {
        return new DestinationPatchConfig\FifoEndpoint($fifoEndpoint);
    }

    public static function amazonS3(S3ConfigPatch $amazonS3): DestinationPatchConfig\AmazonS3
    {
        return new DestinationPatchConfig\AmazonS3($amazonS3);
    }

    public static function googleCloudStorage(GoogleCloudStorageConfigPatch $googleCloudStorage): DestinationPatchConfig\GoogleCloudStorage
    {
        return new DestinationPatchConfig\GoogleCloudStorage($googleCloudStorage);
    }

    public static function googleCloudPubSub(GoogleCloudPubSubConfigPatch $googleCloudPubSub): DestinationPatchConfig\GoogleCloudPubSub
    {
        return new DestinationPatchConfig\GoogleCloudPubSub($googleCloudPubSub);
    }

    public static function sqs(SqsConfigPatch $sqs): DestinationPatchConfig\Sqs
    {
        return new DestinationPatchConfig\Sqs($sqs);
    }

    public static function sns(SnsConfigPatch $sns): DestinationPatchConfig\Sns
    {
        return new DestinationPatchConfig\Sns($sns);
    }

    public static function bigQuery(BigQueryConfigPatch $bigQuery): DestinationPatchConfig\BigQuery
    {
        return new DestinationPatchConfig\BigQuery($bigQuery);
    }

    public static function clickhouse(ClickhouseConfigPatch $clickhouse): DestinationPatchConfig\Clickhouse
    {
        return new DestinationPatchConfig\Clickhouse($clickhouse);
    }

    public static function eventBridge(EventBridgeConfigPatch $eventBridge): DestinationPatchConfig\EventBridge
    {
        return new DestinationPatchConfig\EventBridge($eventBridge);
    }

    public static function snowflake(SnowflakeConfigPatch $snowflake): DestinationPatchConfig\Snowflake
    {
        return new DestinationPatchConfig\Snowflake($snowflake);
    }

    public static function rabbitMq(RabbitMqConfigPatch $rabbitMq): DestinationPatchConfig\RabbitMq
    {
        return new DestinationPatchConfig\RabbitMq($rabbitMq);
    }

    public static function redshift(RedshiftConfigPatch $redshift): DestinationPatchConfig\Redshift
    {
        return new DestinationPatchConfig\Redshift($redshift);
    }

    public static function postgres(PostgresConfigPatch $postgres): DestinationPatchConfig\Postgres
    {
        return new DestinationPatchConfig\Postgres($postgres);
    }

    public static function fromTypeAndConfig(string $type, mixed $config): self
    {
        $config ??= [];

        return match ($type) {
            'pollingEndpoint' => self::pollingEndpoint(),

            'azureBlobStorage' => self::azureBlobStorage(AzureBlobStorageConfigPatch::fromMixed($config)),

            'otelTracing' => self::otelTracing(OtelTracingConfigPatch::fromMixed($config)),

            'fifoEndpoint' => self::fifoEndpoint(SinkHttpConfigPatch::fromMixed($config)),

            'amazonS3' => self::amazonS3(S3ConfigPatch::fromMixed($config)),

            'googleCloudStorage' => self::googleCloudStorage(GoogleCloudStorageConfigPatch::fromMixed($config)),

            'googleCloudPubSub' => self::googleCloudPubSub(GoogleCloudPubSubConfigPatch::fromMixed($config)),

            'sqs' => self::sqs(SqsConfigPatch::fromMixed($config)),

            'sns' => self::sns(SnsConfigPatch::fromMixed($config)),

            'bigQuery' => self::bigQuery(BigQueryConfigPatch::fromMixed($config)),

            'clickhouse' => self::clickhouse(ClickhouseConfigPatch::fromMixed($config)),

            'eventBridge' => self::eventBridge(EventBridgeConfigPatch::fromMixed($config)),

            'snowflake' => self::snowflake(SnowflakeConfigPatch::fromMixed($config)),

            'rabbitMq' => self::rabbitMq(RabbitMqConfigPatch::fromMixed($config)),

            'redshift' => self::redshift(RedshiftConfigPatch::fromMixed($config)),

            'postgres' => self::postgres(PostgresConfigPatch::fromMixed($config)),

            default => throw new \InvalidArgumentException("Unknown type: {$type}"),
        };
    }

    /**
     * Create an instance from a mixed obj.
     */
    public static function fromMixed(mixed $data): self
    {
        return self::fromTypeAndConfig(
            \Svix\Utils::deserializeString($data, 'type', true, 'DestinationPatchConfig'),
            \Svix\Utils::getValFromJson($data, 'config', false, 'DestinationPatchConfig') ?? []
        );
    }

    /**
     * Create an instance from a json string.
     */
    public static function fromJson(string $json): self
    {
        $data = json_decode(json: $json, associative: true, depth: 512, flags: JSON_THROW_ON_ERROR);

        return self::fromMixed($data);
    }

    public function jsonSerialize(): mixed
    {
        return \Svix\Utils::newStdClassIfArrayIsEmpty($this->configPayload());
    }
}

namespace Svix\Models\DestinationPatchConfig;

final class PollingEndpoint extends \Svix\Models\DestinationPatchConfig
{
    public function variantName(): string
    {
        return 'pollingEndpoint';
    }

    public function configPayload(): mixed
    {
        return new \stdClass();
    }
}

final class AzureBlobStorage extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\AzureBlobStorageConfigPatch $azureBlobStorage)
    {
    }

    public function variantName(): string
    {
        return 'azureBlobStorage';
    }

    public function configPayload(): mixed
    {
        return $this->azureBlobStorage;
    }
}

final class OtelTracing extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\OtelTracingConfigPatch $otelTracing)
    {
    }

    public function variantName(): string
    {
        return 'otelTracing';
    }

    public function configPayload(): mixed
    {
        return $this->otelTracing;
    }
}

final class FifoEndpoint extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\SinkHttpConfigPatch $fifoEndpoint)
    {
    }

    public function variantName(): string
    {
        return 'fifoEndpoint';
    }

    public function configPayload(): mixed
    {
        return $this->fifoEndpoint;
    }
}

final class AmazonS3 extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\S3ConfigPatch $amazonS3)
    {
    }

    public function variantName(): string
    {
        return 'amazonS3';
    }

    public function configPayload(): mixed
    {
        return $this->amazonS3;
    }
}

final class GoogleCloudStorage extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\GoogleCloudStorageConfigPatch $googleCloudStorage)
    {
    }

    public function variantName(): string
    {
        return 'googleCloudStorage';
    }

    public function configPayload(): mixed
    {
        return $this->googleCloudStorage;
    }
}

final class GoogleCloudPubSub extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\GoogleCloudPubSubConfigPatch $googleCloudPubSub)
    {
    }

    public function variantName(): string
    {
        return 'googleCloudPubSub';
    }

    public function configPayload(): mixed
    {
        return $this->googleCloudPubSub;
    }
}

final class Sqs extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\SqsConfigPatch $sqs)
    {
    }

    public function variantName(): string
    {
        return 'sqs';
    }

    public function configPayload(): mixed
    {
        return $this->sqs;
    }
}

final class Sns extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\SnsConfigPatch $sns)
    {
    }

    public function variantName(): string
    {
        return 'sns';
    }

    public function configPayload(): mixed
    {
        return $this->sns;
    }
}

final class BigQuery extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\BigQueryConfigPatch $bigQuery)
    {
    }

    public function variantName(): string
    {
        return 'bigQuery';
    }

    public function configPayload(): mixed
    {
        return $this->bigQuery;
    }
}

final class Clickhouse extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\ClickhouseConfigPatch $clickhouse)
    {
    }

    public function variantName(): string
    {
        return 'clickhouse';
    }

    public function configPayload(): mixed
    {
        return $this->clickhouse;
    }
}

final class EventBridge extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\EventBridgeConfigPatch $eventBridge)
    {
    }

    public function variantName(): string
    {
        return 'eventBridge';
    }

    public function configPayload(): mixed
    {
        return $this->eventBridge;
    }
}

final class Snowflake extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\SnowflakeConfigPatch $snowflake)
    {
    }

    public function variantName(): string
    {
        return 'snowflake';
    }

    public function configPayload(): mixed
    {
        return $this->snowflake;
    }
}

final class RabbitMq extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\RabbitMqConfigPatch $rabbitMq)
    {
    }

    public function variantName(): string
    {
        return 'rabbitMq';
    }

    public function configPayload(): mixed
    {
        return $this->rabbitMq;
    }
}

final class Redshift extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\RedshiftConfigPatch $redshift)
    {
    }

    public function variantName(): string
    {
        return 'redshift';
    }

    public function configPayload(): mixed
    {
        return $this->redshift;
    }
}

final class Postgres extends \Svix\Models\DestinationPatchConfig
{
    public function __construct(public readonly \Svix\Models\PostgresConfigPatch $postgres)
    {
    }

    public function variantName(): string
    {
        return 'postgres';
    }

    public function configPayload(): mixed
    {
        return $this->postgres;
    }
}
